plugins/vpp/dnsplugin/descriptor/dnsserver.go

```diff
@@ -0,0 +1,149 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package descriptor
+
+import (
+	"net"
+
+	"github.com/go-errors/errors"
+	"go.ligato.io/cn-infra/v2/logging"
+	scheduler "go.ligato.io/vpp-agent/v3/plugins/kvscheduler/api"
+	"go.ligato.io/vpp-agent/v3/plugins/vpp/dnsplugin/descriptor/adapter"
+	"go.ligato.io/vpp-agent/v3/plugins/vpp/dnsplugin/vppcalls"
+	dns "go.ligato.io/vpp-agent/v3/proto/ligato/vpp/dns"
+)
+
+const (
+	// DNSServerDescriptorName is the name of the descriptor for VPP DNS server functionality
+	DNSServerDescriptorName = "vpp-dns-server"
+)
+
+// DNSServerDescriptor teaches KVScheduler how to configure VPP to act as DNS server.
+type DNSServerDescriptor struct {
+	// dependencies
+	log        logging.Logger
+	dnsHandler vppcalls.DNSVppAPI
+}
+
+// NewDNServerDescriptor creates a new instance of the DNSServer descriptor.
+func NewDNServerDescriptor(dnsHandler vppcalls.DNSVppAPI, log logging.PluginLogger) *scheduler.KVDescriptor {
+	ctx := &DNSServerDescriptor{
+		log:        log.NewLogger("dnsserver-descriptor"),
+		dnsHandler: dnsHandler,
+	}
+
+	typedDescr := &adapter.DNSServerDescriptor{
+		Name:            DNSServerDescriptorName,
+		KeySelector:     dns.ModelDNSServer.IsKeyValid,
+		ValueTypeName:   dns.ModelDNSServer.ProtoName(),
+		KeyLabel:        dns.ModelDNSServer.StripKeyPrefix,
+		NBKeyPrefix:     dns.ModelDNSServer.KeyPrefix(),
+		Validate:        ctx.ValidateDNSServers,
+		ValueComparator: ctx.EquivalentDNSServers,
+		Create:          ctx.Create,
+		Delete:          ctx.Delete,
+		Update:          ctx.Update,
+	}
+	return adapter.NewDNSServerDescriptor(typedDescr)
+}
+
+// ValidateDNSServers validates content of DNS server configuration
+func (d *DNSServerDescriptor) ValidateDNSServers(key string, dnsServer *dns.DNSServer) error {
+	if len(dnsServer.UpstreamDnsServers) == 0 {
+		return scheduler.NewInvalidValueError(
+			errors.New("at least one upstream DNS server must be defined"), "upstreamDnsServers")
+	}
+	for _, serverIpAddress := range dnsServer.UpstreamDnsServers {
+		if net.ParseIP(serverIpAddress) == nil {
+			return scheduler.NewInvalidValueError(errors.Errorf("failed to parse upstream DNS Server IP "+
+				"address %s, should be a valid ipv4/ipv6 address", serverIpAddress), "upstreamDnsServers")
+		}
+	}
+	return nil
+}
+
+// EquivalentDNSServers determines whether 2 DNS servers are logically equal. This comparison takes
+// into consideration also semantics that couldn't be modeled into proto models (i.e. upstream servers are
+// IP addresses and not only strings)
+func (d *DNSServerDescriptor) EquivalentDNSServers(key string, oldDNSServer, newDNSServer *dns.DNSServer) bool {
```

Comment:
EquivalentDNSServers could compare UpstreamDnsServers as sets, i.e. different order of the same set of values could be considered equivalent. Otherwise it behaves just like proto.Equal and so it is not neccessarry.

Reply:
Upstream DNS server is a list and not a set. Removing EquivalentDNSServers. Fixed in [`ac77bf2`](https://github.com/ligato/vpp-agent/pull/1779/commits/ac77bf2faa7c4b796216c7917a545284c73d97d3)